plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -763,13 +764,30 @@ private Optional<ConnectorTableLayout> getWriteLayout(Schema tableSchema, Partit
             return Optional.empty();
         }
 
-        validateNotPartitionedByNestedField(tableSchema, partitionSpec);
-        Map<Integer, IcebergColumnHandle> columnById = getColumns(tableSchema, typeManager).stream()
+        Map<Integer, Integer> indexParents = indexParents(tableSchema.asStruct());
+        Map<Integer, IcebergColumnHandle> columnById = getProjectedColumns(tableSchema, typeManager).stream()
                 .collect(toImmutableMap(IcebergColumnHandle::getId, identity()));
 
         List<IcebergColumnHandle> partitioningColumns = partitionSpec.fields().stream()
                 .sorted(Comparator.comparing(PartitionField::sourceId))
-                .map(field -> requireNonNull(columnById.get(field.sourceId()), () -> "Cannot find source column for partitioning field " + field))
+                .map(field -> {
+                    boolean isBaseColumn = !indexParents.containsKey(field.sourceId());
+                    int sourceId;
+                    if (isBaseColumn) {
+                        sourceId = field.sourceId();
+                    }
+                    else {
+                        sourceId = getParentSourceId(indexParents, field.sourceId());
+                    }
+                    Type sourceType = tableSchema.findType(sourceId);
+                    if (sourceType.isMapType()) {
+                        throw new TrinoException(NOT_SUPPORTED, "Partitioning field [" + field.name() + "] cannot be contained in a map");
+                    }
+                    if (sourceType.isListType()) {
+                        throw new TrinoException(NOT_SUPPORTED, "Partitioning field [" + field.name() + "] cannot be contained in a array");
+                    }
```

Comment:
Please update the PR description and commit message to say that it adds support for partitioning on ROW fields specifically as "nested" would include map, array and combinations of row/map/array.
Also, maybe add a code comment somewhere which points to iceberg spec about what nested types are supported as a partitioning field.